lib/streamlit/elements/multiselect.py

```diff
@@ -264,6 +291,20 @@ def _multiselect(
             serializer=serde.serialize,
             ctx=ctx,
         )
+
+        if (
+            max_selections is not None
+            and get_default_count(widget_state.value) > max_selections
+        ):
+            raise StreamlitAPIException(
+                f"""           
+Multiselect has {get_default_count(widget_state.value)} options selected but `max_selections` 
+is set to {max_selections}. This happened because you manipulated 
+the widget's state through `st.session_state`. Note that 
+this happened before the line indicated in the traceback. 
+Please select at most {max_selections} options.
+"""
+            )
```

Comment:
Is there a way of combining this check with the one above? For example, if `default` is set to have more options than `max_selections` and we didn't have the check above, wouldn't we eventually run into it here?

It seems a bit iffy to have two more or less structurally identical checks that differ only in the exact error message that's printed out. On the product side, we might want this, but it feels clunky to have duplicate code when we could probably just have a general error message mentioning both possibilities

Reply:
Hm so we want to catch the max_selections error before registering the widget. In addition, we want to check after registering the widget for session state specifically so I'm not sure if the code can really change. 

I think it's better from a product perspective to give clearer messages and not have a general message even though it does make the code a little clunky. As Snowflake says, everything is about the customer so I think it's best to do it like this.

Reply:
What's the difference in user-facing behavior if we throw the error before or after `register_widget`? I don't quite understand if there's any obvious difference between the two for a developer since the script terminates with a `StreamlitAPIException` either way.

I also don't think the product experience degrades very much having a more general error message that makes it worth making the codebase worse to support, but that's ultimately a product decision. We could also, if we really have to keep both messages, at least write a helper function so that we don't have to duplicate structurally identical code

Reply:
SG let's create one general error message for both situations with:

```
Multiselect has {get_default_count(widget_state.value)} options selected but `max_selections` 
is set to {max_selections}. This happened because you either gave too many options to `default` or you manipulated the widget's state through `st.session_state`. Note that 
the latter can happen before the line indicated in the traceback. 
Please select at most {max_selections} options.
```

Reply:
(need to clean up linebreaks)

Reply:
Cool fixed the messaging! Thanks!